### PR TITLE
Add grid length

### PR DESCRIPTION
**Pull request by adriendelsalle**
Add a `length` method to get the physical length of the grid in all its dimensions.

Related to #59.

Details:
- add factory method `from_length`
- add tests for profile and raster grids
- fix ICE (internal compiler error) on MSVC14.0 (VS2015)

- [x] Rebase after #57 merged

### CODE REVIEW

**Comment by adriendelsalle** [comment]
Rebased done! Ready for review @benbovy

File: include/fastscapelib/profile_grid.hpp [hunk 1]
```diff
@@ -207,19 +217,43 @@ namespace fastscapelib
...
-                                        spacing_type spacing,
-                                        const boundary_status_type& status_at_bounds,
-                                        const std::vector<node>& status_at_nodes)
+                                            spacing_type spacing,
+                                            const boundary_status_type& status_at_bounds,
+                                            const std::vector<node>& status_at_nodes)
         : base_type(size), m_size(size), m_spacing(spacing), m_status_at_bounds(status_at_bounds)
     {
         m_shape = {static_cast<typename shape_type::value_type>(m_size)};
+        m_length = static_cast<spacing_type>(size) * spacing;
```
Comment (benbovy) [inline include/fastscapelib/profile_grid.hpp:226]:
```suggestion
        m_length = static_cast<spacing_type>(size - 1) * spacing;
```

File: include/fastscapelib/profile_grid.hpp [hunk 2]
```diff
@@ -207,19 +217,43 @@ namespace fastscapelib
...
+     * @param status_at_bounds Status at boundary nodes (left & right grid edges).
+     * @param status_at_nodes Manually define the status at any node on the grid.
+     */
+    template <class XT, class C>
+    profile_grid_xt<XT, C> profile_grid_xt<XT, C>::from_length(size_type size,
+                                                               length_type length,
+                                                               const boundary_status_type& status_at_bounds,
+                                                               const std::vector<node>& status_at_nodes)
+    {
+        spacing_type spacing = length / static_cast<length_type>(size);
```
Comment (benbovy) [inline include/fastscapelib/profile_grid.hpp:252]:
```suggestion
        spacing_type spacing = length / static_cast<length_type>(size - 1);
```

File: include/fastscapelib/raster_grid.hpp [hunk 3]
```diff
@@ -281,14 +289,14 @@ namespace fastscapelib
...
-                                       const boundary_status_type& status_at_bounds,
-                                       const std::vector<raster_node>& status_at_nodes)
+    raster_grid_xt<XT, C>::raster_grid_xt(const shape_type& shape,
+                                          const spacing_type& spacing,
+                                          const boundary_status_type& status_at_bounds,
+                                          const std::vector<raster_node>& status_at_nodes)
         : base_type(shape[0] * shape[1]), m_shape(shape), m_spacing(spacing), m_status_at_bounds(status_at_bounds)
     {
         m_size = shape[0] * shape[1];
+        m_length = xt::adapt(shape) * spacing;
```
Comment (benbovy) [inline include/fastscapelib/raster_grid.hpp:299]:
```suggestion
        m_length = (xt::adapt(shape) - 1) * spacing;
```

File: include/fastscapelib/raster_grid.hpp [hunk 4]
```diff
@@ -303,6 +311,29 @@ namespace fastscapelib
...
+     * @param status_at_bounds Status at boundary nodes (left & right grid edges).
+     * @param status_at_nodes Manually define the status at any node on the grid.
+     */
+    template <class XT, class C>
+    raster_grid_xt<XT, C> raster_grid_xt<XT, C>::from_length(const shape_type& shape,
+                                                             const length_type& length,
+                                                             const boundary_status_type& status_at_bounds,
+                                                             const std::vector<raster_node>& status_at_nodes)
+    {
+        spacing_type spacing = length / xt::adapt(shape);
```
Comment (benbovy) [inline include/fastscapelib/raster_grid.hpp:332]:
```suggestion
        spacing_type spacing = length / (xt::adapt(shape) - 1);
```

File: python/src/grid.cpp [hunk 5]
```diff
@@ -54,19 +54,23 @@ void add_grid_bindings(py::module& m) {
...
+                {
+                    std::vector<fs::node> node_vec;
+                    for (auto&& node : ns)
+                    {
+                        node_vec.push_back({node.first, node.second});
+                    }
+                    return std::make_unique<profile_grid>(size, spacing, bs, node_vec);
+                }));
+
+    pgrid.def_static("from_length", &profile_grid::from_length);
```
Comment (benbovy) [inline python/src/grid.cpp:69]: Perhaps we could [bind this factory as a constructor](https://pybind11.readthedocs.io/en/stable/advanced/classes.html?highlight=factory#custom-constructors), which is a bit more Python idiomatic.

File: python/src/grid.cpp [hunk 6]
```diff
@@ -101,10 +105,20 @@ void add_grid_bindings(py::module& m) {
 
     // ==== Binding of the RasterGrid class ==== //
     py::class_<raster_grid> rgrid(grid_m, "RasterGrid");
-    rgrid.def(py::init<std::array<std::size_t, 2>, const raster_grid::spacing_type, const raster_grid::boundary_status_type&, const std::vector<fs::raster_node>>());
+    rgrid.def(py::init<const raster_grid::shape_type&, const xt::pytensor<double, 1>&, const raster_grid::boundary_status_type&, const std::vector<fs::raster_node>>());
+
+    rgrid.def_static("from_length", [](const raster_grid::shape_type& shape,
```
Comment (benbovy) [inline python/src/grid.cpp:110]: Same suggestion than in my comment above.

**Review by benbovy — COMMENTED** [review]
Thanks @adriendelsalle! I've left a few suggestions to fix the calculation of the total grid length and regarding the Python bindings.

File: python/src/grid.cpp (hunk 5 above)
Comment (adriendelsalle) [reply python/src/grid.cpp:69]: not sure it's possible, for the same reason we needed to add a static method with a factory pattern : the constructor and this method have the same signature.
I will try :)

File: include/fastscapelib/profile_grid.hpp (hunk 1 above)
Comment (adriendelsalle) [reply include/fastscapelib/profile_grid.hpp:226]: looks like a typo... I will modify tests to be able to capture this! thanks!

File: python/src/grid.cpp (hunk 5 above)
Comment (benbovy) [reply python/src/grid.cpp:69]: Ah, right. No big deal if that's not possible.

Comment (adriendelsalle) [reply python/src/grid.cpp:69]: There is no compilation error but only the regular cpp constructor is used.

**Comment by adriendelsalle** [comment]
I've fixed the implementation, good catch ! and sorry for that.

Concerning the Python bindings, it's possible to bind multiple factories/constructors as a Python constructor only when they provide different signature, which is not the case here. It is also the reason why we had to implement a factory pattern `from_length`.

I will squash commits before merging if it's ok for you now.

**Review by benbovy — APPROVED** [review]
> Concerning the Python bindings, it's possible to bind multiple factories/constructors as a Python constructor only when they provide different signatures, which is not the case here. It is also the reason why we had to implement a factory pattern from_length.

Yes that's right. Thinking more about that, it's probably better if the API of the Python bindings do not differ too much from the C++ API, even if that's a bit less idiomatic. It will be less confusing and easier to maintain (e.g., doc). 

Thanks @adriendelsalle !

**Comment by benbovy** [comment]
> I will squash commits before merging if it's ok for you now.

No need for you to squash the commits, I use GitHub's "Squash and merge" for the PRs here.

**Pull request merged**